csrc/scheduler/multi_matmul.cpp

```diff
@@ -0,0 +1,421 @@
+          cache_op = vec_bytes == 16LL ? CacheOp::Global : CacheOp::AllLevels;
+        };
+
+        NVF_ERROR(operand->uses().size() == 1);
+        smem_operands[i] = ir_utils::consumerTvsOf(operand).at(0);
+        smem_operands[i]->definition()->as<LoadStoreOp>()->setOpType(load_op);
+        smem_operands[i]->definition()->as<LoadStoreOp>()->setCacheOp(cache_op);
+        if (smem_operands[i]->uses().size() > 1) {
+          // There can be multiple uses for example if we have A @ B1 + A @ B2
+          // then A will be cached to smem then it might be loaded into two
+          // separate register buffers, one for each mma. Instead, we will load
+          // it once into registers then re-use the register buffer for both
+          // mmas.
+          cacheAfter(smem_operands[i]);
+        }
+        NVF_ERROR(smem_operands[i]->uses().size() == 1);
+        smem_operands[i]->setMemoryType(MemoryType::Shared);
+      }
+    };
+    cacheOperandsToSmem(as_, acw_smems_, params_.supported_vec_size.a);
+    cacheOperandsToSmem(bs_, bcw_smems_, params_.supported_vec_size.b);
+
+    // We add two LoadStore operators to the inputs of our fusions. The first
+    // one is for a read from global memory and the second one (below) is for a
+    // cache read. As an optimizaton, we avoid adding an operator if there's an
+    // existing LoadStoreOp present. Please note that for the second LoadStore
+    // we don't propagate the allocation domain, since the scheduler sets the
+    // allocation domain in the registers.
+    auto addSetsForCacheReads = [&](const std::vector<TensorView*>& tv_smems,
+                                    std::vector<TensorView*>& tv_rs) {
+      tv_rs.resize(tv_smems.size(), nullptr);
+      for (size_t i : c10::irange(tv_smems.size())) {
+        TensorView* tv_smem = tv_smems[i];
+        TensorView*& tv_r = tv_rs[i];
+
+        if (auto ldst = dynamic_cast<LoadStoreOp*>(tv_smem->uses().at(0))) {
+          tv_r = ldst->out()->as<TensorView>();
+          ldst->setOpType(LoadStoreOpType::LdMatrix);
+        } else {
+          tv_r = cacheAfter(
+              tv_smem,
+              LoadStoreOpType::LdMatrix,
+              CacheOp::Unspecified,
+              /*propagate_allocation_domain=*/false);
+        }
+      }
+    };
+    // Shared memory read
+    addSetsForCacheReads(acw_smems_, acrs_);
+    addSetsForCacheReads(bcw_smems_, bcrs_);
+  }
+
+  //! This calls orig->cacheAfter() and also updates the permissive graph to
+  //! reflect the new IterDomain mappings
+  TensorView* cacheAfter(
+      TensorView* orig,
+      LoadStoreOpType op_type = LoadStoreOpType::Set,
+      CacheOp cache_op = CacheOp::AllLevels,
+      bool propagate_allocation_domain = false) {
+    const std::vector<IterDomain*> orig_alloc =
+        orig->getMaybeAllocationDomain();
+
+    TensorView* c =
+        orig->cacheAfter(op_type, cache_op, propagate_allocation_domain);
+
+    if (propagate_allocation_domain) {
```

Comment:
Why not always add allocation domain to ValGraph? Or, maybe better, could we just add `tv->domain()->allIDs()`?

Reply:
Yeah that's a good point. I figured if we didn't propagate allocation domain then it would be empty anyway. But it could contain transforms so we should initialize the whole path not just the IDs in the alloc domain itself. I _think_ we should have the same ordering of IDs when using `orig->allIDs()` and `c->allIDs()`. I need that correspondence in order to add them to the correct group. I will give this a try (replacing both these loops with one over `allIDs`) with a note about the correspondence.